sf-stream/src/runtime.rs

```diff
@@ -0,0 +1,227 @@
+    pub block_height: u64,
+    pub current_epoch: u64,
+    // This is only ever used for testing
+    pub mp_sender: MempoolClientSender,
+}
+
+impl SfStreamer {
+    pub fn new(
+        context: Arc<Context>,
+        starting_version: u64,
+        mp_client_sender: Option<MempoolClientSender>,
+    ) -> Self {
+        let resolver = Arc::new(context.move_resolver().unwrap());
+        let latest = context.get_latest_ledger_info().unwrap();
+        let block_info = context
+            .get_block_info(starting_version, latest.ledger_version.0)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Could not get block_info for starting version {} and ledger_version {}",
+                    starting_version, latest.ledger_version.0
+                )
+            });
+        let starting_tnx = context
+            .get_transaction_by_version(block_info.start_version, latest.ledger_version.0)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Could not get starting_tnx for starting version {} and ledger_version {}",
+                    starting_version, latest.ledger_version.0
+                )
+            });
+
+        let (version, epoch) = match starting_tnx.transaction {
+            aptos_types::transaction::Transaction::BlockMetadata(bmt) => {
+                (starting_tnx.version, bmt.epoch())
+            }
+            aptos_types::transaction::Transaction::GenesisTransaction(_gt) => (0, 0),
+            _ => panic!(
+                "[sf-stream] first transaction is not a block metadata or genesis transaction"
+            ),
+        };
+
+        // fake mempool client/sender, if we need to, so we can use the same code for both api and sf-streamer
+        let mp_client_sender = mp_client_sender.unwrap_or_else(|| {
+            let (mp_client_sender, _mp_client_events) = channel(1);
+            mp_client_sender
+        });
+
+        Self {
+            context,
+            current_version: version,
+            resolver,
+            block_height: block_info.block_height,
+            current_epoch: epoch,
+            mp_sender: mp_client_sender,
+        }
+    }
+
+    pub async fn start(&mut self) {
+        loop {
+            self.batch_convert_once(100).await;
+        }
+    }
+
+    pub async fn batch_convert_once(&mut self, batch_size: u16) -> Vec<extractor::Transaction> {
+        let mut result: Vec<extractor::Transaction> = vec![];
+        match &self.context.db.get_first_txn_version() {
+            Ok(version_result) => {
+                if let Some(oldest_version) = version_result {
+                    if oldest_version > &self.current_version {
+                        warn!(
+                            "[sf-stream] oldest txn version is {} but requested version is {}",
+                            oldest_version, &self.current_version
+                        );
+                        sleep(Duration::from_millis(300)).await;
```

Comment:
I'm wondering where are all the magic sleep numbers coming from?  These should be adjustable